### DISCUSSION TITLE
SBOM Author in SBOM

### DISCUSSION THREAD

**Discussion by MacOS** [oss-review-toolkit/ort#9003] [Q&A]
Hi all,

I have a question pertaining to SBOM.

Can ORT automatically add the SBOM author information based on some config? This is imporant because it is a required SBOM field as per [BSI-TR-03183-2](https://www.bsi.bund.de/SharedDocs/Downloads/EN/BSI/Publications/TechGuidelines/TR03183/BSI-TR-03183-2.pdf?__blob=publicationFile&v=5) - search in the document for *Creator of the SBOM*. **Please note that it should be possible to add an E-Mail address or at least a URL**. In CycloneDX 1.6, this relates to [the authors field in the metadata](https://cyclonedx.org/docs/1.6/json/#metadata_authors). In SPDX 2.3, this relates to the [*creators* field within *creationInfo*](https://github.com/spdx/spdx-spec/blob/ecacc878b146c0bf38028397514615abb372e0e8/examples/SPDXJSONExample-v2.3.spdx.json#L4-L7).

many thanks.

**Comment by sschuberth** [comment]
Thanks for bringing up those questions! For CycloneDX, we currently do not set the `authors` field at all, but that should be very easy to add via a reporter-specific option. For SPDX, there already is [a reporter-specific option to set the `creationInfo.comment`](https://github.com/oss-review-toolkit/ort/blob/88f45487de170e34e36f06f06d81d3211635c53e/plugins/reporters/spdx/src/main/kotlin/SpdxDocumentReporter.kt#L44), but I guess the author should go as a `Person: ` entry into `creationInfo.creators`, right?

**Reply by MacOS** [reply]
Thank you for the fast reply, @sschuberth!

@CycloneDX
I will try that out as soon as I can.

@SPDX
Yes, it should go there.

**Reply by MacOS** [reply]
I just looked at the CycloneDX reporter code here.

https://github.com/oss-review-toolkit/ort/blob/88f45487de170e34e36f06f06d81d3211635c53e/plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt#L64-L69

I do not understand how any of these CyclonceDX specifix options help me in setting the author inormation. May I ask you to be more specific?

Thank you in advance!

**Reply by sschuberth** [reply]
> but that should be very easy to add via a reporter-specific option. 

By "add" here I was meaning that such a reporter-specific option still needs to be added *in the code*, i.e. the feature is not there yet 😉

**Reply by MacOS** [reply]
Oh, I see - sorry for the missunderstanding.

Wouldn't the SPDX reporter also need a code change? Or am I miss-reading the code?

In my opinion, it would be need if one could set that within the repository specific ORT configuration file. Also in my opinion, it would make sense to set this information globally because as a user, I do not want to add the information separately for each report.

**Reply by MacOS** [reply]
On a side note, I'm happy to take over the task of writting the feature requests, and also contributing the code itself. :)

**Reply by sschuberth** [reply]
> Wouldn't the SPDX reporter also need a code change? Or am I miss-reading the code?

Now that you confirmed that the author should go to `creationInfo.creators` (and not `creationInfo.comment`, which was already configurable), yes 😄 

> In my opinion, it would be need if one could set that within the repository specific ORT configuration file. Also in my opinion, it would make sense to set this information globally because as a user, I do not want to add the information separately for each report.

I tend to agree; however, both would be bigger changes because reporter-specific options are not plugin options yet (see the somewhat related https://github.com/oss-review-toolkit/ort/issues/6507).

> On a side note, I'm happy to take over the task of writting the feature requests, and also contributing the code itself. :)

Many thanks! For SPDX, I just finished the necessary code change 😁 See https://github.com/oss-review-toolkit/ort/pull/9004.

**Reply by MacOS** [reply]
Ok, this was fast.

Do I understand your code changes in #9004 correctly, that I can add the following information
```
"Person: FirstName LastName (first_name.last_name@my_organisation.com)"
"Organization: MyOrganisation (opensource@my_organisation.com)"
```
with
```bash
./ort/cli/build/install/ort/bin/ort \
    report \
        ....
        --report-format=SpdxDocument \
        --report-option=SpdxDocument=creationInfo.person="Person: FirstName LastName (first_name.last_name@my_organisation.com)" \
        --report-option=SpdxDocument=creationInfo.organization="Organization: MyOrganisation (opensource@my_organisation.com)"
```
and I will get
```xml
<creationInfo>
.....
    <creators>Organization: MyOrganisation (opensource@my_organisation.com)</creators>
    <creators>Person: FirstName LastName (first_name.last_name@my_organisation.com)</creators>
</creationInfo>
```

Can I also set these with a repository specific ORT file?

**Reply by sschuberth** [reply]
> Do I understand your code changes in https://github.com/oss-review-toolkit/ort/pull/9004 correctly

Yes. But now that you write it out, my intention of distinguishing between `creationInfo.person` and `creationInfo.organization` was to set the necessary prefixes internally. I'll amend my change.

> Can I also set these with a repository specific ORT file?

No, that's not possible yet (for any reporter), and would be a bigger change.

**Reply by MacOS** [reply]
> Yes. But now that you write it out, my intention of distinguishing between creationInfo.person and creationInfo.organization was to set the necessary prefixes internally. I'll amend my change.

Thank you!

> No, that's not possible yet (for any reporter), and would be a bigger change.

Out of interest, what would be these changes?

**Reply by sschuberth** [reply]
> Out of interest, what would be these changes?

Turns out I was partly mislead: ORT already [can read reporter-specific options from the global `config.yml`](https://github.com/oss-review-toolkit/ort/blob/88f45487de170e34e36f06f06d81d3211635c53e/plugins/commands/reporter/src/main/kotlin/ReporterCommand.kt#L271-L274), but such options cannot currently be overridden by the repository-specific `.ort.yml`.

**Reply by MacOS** [reply]
That is great!

I still suspect that there is a common use case where organisations want to set the author, and other metadata, on a repository-specific ``.ort.yml``

**Reply by sschuberth** [reply]
> I still suspect that there is a common use case where organisations want to set the author, and other metadata, on a repository-specific .ort.yml

Probably. Please file a feature request issue for it to track the ask.

**Reply by MacOS** [reply]
I will!

Thank you again for the fast responses to my questions and remarks, as well as the fast pull request. I highly appreciate that!
